File: day5/src/main.py
def get_input(filename):
    with open(filename, 'r') as f:
        text_lines = f.readlines()
        out = [[tuple(int(s) for s in p.split(','))
               for p in text_line.split(" -> ")]
               for text_line in text_lines]

    return out


def is_horizontal(line):
    if line[0][1] == line[1][1]:
        return True
    return False


def get_horizontal(lines):
    out = []
    for line in lines:
        if is_horizontal(line):
            out.append(line)
    return out

    def print_board(board):
        for l in board:
            print(l)


def part1(lines):
    board = [[0]*10]*10
    print("Initial board: ")
    print_board(board)

    for line in lines:
        if is_horizontal(line):
            if line[0][0] > line[1][0]:
                startpoint = line[1][0]
                endpoint = line[0][0]
            else:
                startpoint = line[0][0]
                endpoint = line[1][0]
            for i in range(startpoint, endpoint + 1):
                board[i][line[0][1]] += 1

    print_board(board)


if __name__ == '__main__':
    filename = "day5_sample.in"
    res = get_input(filename)
    horizontal = get_horizontal(res)
    print(horizontal)

File: day5/src/main.rs
mod line;
mod point;

use line::Line;

use std::fs;

const BOARD_SIZE: usize = 1000;

fn main() {
    // Read input
    let filename = "day5.in";
    let data: Vec<String> = match fs::read_to_string(filename) {
        Ok(data) => data.split('\n').map(|x| x.to_string()).collect(),
        Err(why) => panic!("Couldn't read file {}", why),
    };
    let mut lines: Vec<Line> = Vec::new();
    for d in data.into_iter() {
        lines.push(Line::from(d));
    }
    let mut max_dim = 0;

    for line in &lines {
        max_dim = [max_dim, line.start().x(), line.start().y(), line.end().x(), line.end().y()].into_iter().max().unwrap();
    }

    let mut board = [[0_u32; BOARD_SIZE]; BOARD_SIZE];
    // let mut map: HashMap<(i32, i32), usize> = HashMap::new();

    for line in lines {
        // horizontal -> start.y == end.y
        if line.is_horizontal() {
            let (startpoint, endpoint) = minmax(line.start().x(), line.end().x());
            for i in startpoint..=endpoint {
                board[i as usize][line.start().y() as usize] += 1;
            }
        }

        // vertical -> start.x == end.x
        if line.is_vertical() {
            let (startpoint, endpoint) = minmax(line.start().y(), line.end().y());
            for i in startpoint..=endpoint {
                board[line.start().x() as usize][i as usize] += 1;
            }
        }

        if line.is_diagonal() {
            let line_size = (line.start().x() - line.end().x()).abs();
            let x_direction = match line.start().x() - line.end().x() < 0 {
                true => 1,
                false => -1,
            };
            let y_direction = match line.start().y() - line.end().y() < 0 {
                true => 1,
                false => -1,
            };
            for i in 0..=line_size {
                board[(line.start().x() + i * x_direction) as usize][(line.start().y() + i * y_direction) as usize] += 1;
            }
        }
    }

    let mut sum = 0;
    for (i, b) in board.iter().enumerate() {
        for (j, _) in b.iter().enumerate() {
            if board[i][j] >= 2 {
                sum += 1;
            }
        }
    }
    println!("{}", sum);
}


fn minmax(a: i32, b: i32) -> (i32, i32) {
    if a > b { (b, a) }
    else { (a, b) }
}